### PR TITLE
Document PyObject_SelfIter

### PR DESCRIPTION
**Pull request by hroncok**
Thanks to @encukou for showing me this.

----
📚 Documentation preview 📚: https://cpython-previews--127861.org.readthedocs.build/

### CODE REVIEW

File: Doc/data/refcounts.dat [hunk 1]
```diff
@@ -1849,6 +1849,9 @@ PyObject_RichCompareBool:PyObject*:o1:0:
 PyObject_RichCompareBool:PyObject*:o2:0:
 PyObject_RichCompareBool:int:opid::
 
+PyObject_SelfIter:PyObject*::+1:
+PyObject_SelfIter:PyObject*:obj:0:
```
Comment (hroncok) [inline Doc/data/refcounts.dat:1853]: I want to have the text *Return value: New reference.* show up in the documentation, but I failed. Perhaps this is wrong?

Comment (hroncok) [reply Doc/data/refcounts.dat:1853]: OK, it works on https://cpython-previews--127861.org.readthedocs.build/en/127861/c-api/object.html#c.PyObject_SelfIter -- perhaps a cache thing locally.

**Comment by ZeroIntensity** [comment]
cc @picnixz for the `refcounts.dat` modifications.

File: Doc/c-api/object.rst [hunk 2]
```diff
@@ -509,6 +509,12 @@ Object Protocol
    iterated.
 
 
+.. c:function:: PyObject* PyObject_SelfIter(PyObject *obj)
+
+   This is equivalent to the Python ``__iter__(self): return self`` method.
+   It is intended for iterator types, to be used in the ``tp_iter`` slot.
```
Comment (ZeroIntensity) [inline Doc/c-api/object.rst]: Let's use some references:
```suggestion
   It is intended for :term:`iterator` types, to be used in the :c:member:`PyTypeObject.tp_iter` slot.
```

**Review by vstinner — APPROVED** [review]
LGTM

**Review by StanFromIreland — APPROVED** [review]

**Review by kumaraditya303 — APPROVED** [review]

**Pull request merged**